README.md
RayTracer
=========

TO BE FILLED IN

ChangeLog
2014-11-17 Matt:
	* Adding feature to move x-coordinate

2014-10-07 Matt:
	* Updating how the order of iteration is chosen depending on initial angle.
	* Taggin version 00.03 (for real this time!)

2014-10-03 Matt:
	* Updating main to take option of rotation angle for ice
	* Removing the corner check.  This seems obsolete and was causing some errors.
	* Tagging version 00.03 -- NEVER TAGGED 00.02!  Changing to version 00.02
	* Adding the option to be able to input non-integar steps (so can use floating point)

2014-09-30 Matt:
	* Adding testing file test.C in order to check geometry

2014-09-26 Matt:
	* Added options for running different angles
	* Found some odd bug at angle 110 -- Just stops after two iterations(?)

2014-09-25 Matt:
	* Fixed the bugs due to reflection and refraction at each boundary (I think).
	* Only works for angles 0-90, then it gets confused.  This should be fixed in future!
	* Tagging version.00.00.01
	* Fixed > 90 degree issue.  Now ok for angles 0-180 degree. Had to do with reflections
	* Added reflection method to physics
	* Cleaning up printouts
	* Tagging version.00.00.02
	* For future try to consolidate reflective and refractive methods.

2014-09-24 Matt:
	* Fixed one bug dealing with exiting angle when rotating.
	* Found another issue due with activeSide and the equations list.
	* Consider re-writing now that have better idea of how to handle each case.

2014-09-12 Matt:
	* Committing fix for the 55 angle
	* Still bugs when rotate iceblock. Main issue is refracted angle not in same coord system.

2014-09-11 Matt:
	* Committing working version with a few bugs
	* There is one error that has to do with how you cycle between the sides of the cube.
	  The issue happens at 55 degrees right now, where you get many TIR at the same angle.
	  Fix tomorrow/later

2014-09-10 Matt:
	* RayTracer created. The other programs found were more complicated than needed.
	* Adding basic structure I have thought of so far.
	* Adding skeleton of main to create a window and the ice block.
	* Adding code for the ice block


test.C

#include "../Simulation/scripts/myHist.C"

test()
{

  // Make canvas
  TCanvas* c = makeCanvas("c");
  
  // Make generic hist
  TH1F* hist = makeHist("hist",1,-1,1,"","",kBlack,20);
  hist->SetMinimum(-0.7);
  hist->SetMaximum(0.7);
  hist->Draw();

  // Axis
  TLine* yaxis = makeLine(0,0,-0.5,0.5,kBlack,1); yaxis->Draw("same");

  // Make lines for normal block
  TLine* top   = makeLine(-0.5,0.5,0.15,0.15,kBlue,1);   top->Draw("same");
  TLine* bot   = makeLine(-0.5,0.5,-0.15,-0.15,kBlue,1); bot->Draw("same");
  TLine* right = makeLine(0.5,0.5,-0.15,0.15,kBlue,1);   right->Draw("same");
  TLine* left  = makeLine(-0.5,-0.5,-0.15,0.15,kBlue,1); left->Draw("same");
  
  // Now experiment to try to get this right...
  double angle = -30 * TMath::Pi()/180.;
  //double angle = 0;
  cout<<"tan(angle): "<<tan(angle)<<" "<<-1/tan(angle)<<endl;

  // top
  double x0 = -0.5, x1 = 0.5, y0 = 0.15, y1 = 0.15;
  rotation(angle,x0,y0); rotation(angle,x1,y1);
  cout<<"Top slope: "<<(y1-y0)/(x1-x0)<<endl;
  TLine* ntop = makeLine(x0,x1,y0,y1,kRed,1); ntop->Draw("same");
  
  // bot
  x0 = -0.5; x1 = 0.5; y0 = -0.15; y1 = -0.15;
  rotation(angle,x0,y0); rotation(angle,x1,y1);
  cout<<"Bottom slope: "<<(y1-y0)/(x1-x0)<<endl;
  double slope = (y1-y0)/(x1-x0);
  printEq(slope,x0,y0);
  TLine* ntop = makeLine(x0,x1,y0,y1,kRed,1); ntop->Draw("same");
  
  
  // bot
  x0 = -0.5; x1 = 0.5; y0 = -0.15; y1 = -0.15;
  rotation(angle,x0,y0); rotation(angle,x1,y1);
  slope = (y1-y0)/(x1-x0);
  printEq(slope,x0,y0);
  //cout<<"Bottom slope: "<<(y1-y0)/(x1-x0)<<endl;
  TLine* nbot = makeLine(x0,x1,y0,y1,kRed,1); nbot->Draw("same");

  // right
  x0 = 0.5; x1 = 0.5, y0 = -0.15; y1 = 0.15;
  rotation(angle,x0,y0); rotation(angle,x1,y1);
  cout<<"Right slope: "<<(y1-y0)/(x1-x0)<<endl;
  slope = (y1-y0)/(x1-x0);
  printEq(slope,x0,y0);
  //cout<<"Right slope: "<<(y1-y0)/(x1-x0)<<endl;
  TLine* nR = makeLine(x0,x1,y0,y1,kRed,1); nR->Draw("same");
  
  // left
  x0 = -0.5; x1 = -0.5; y0 = -0.15; y1 = 0.15;
  rotation(angle,x0,y0); rotation(angle,x1,y1);
  cout<<"Left slope: "<<(y1-y0)/(x1-x0)<<endl;
  cout<<"Left slope: "<<1/(y1-y0)/(x1-x0)<<endl;
  TLine* nL = makeLine(x0,x1,y0,y1,kRed,1); nL->Draw("same");


  // Now add something close to what I want to do with
  // the acutal code using theta as the slope
  float slope  = tan(angle);
  float height = 0.3;
  float length = 1.0;
  
  // Top 
  x0 = -length/2., y0 = height/2.;
  evaluate(slope,x1,y0);
  slope = (y1-y0)/(x1-x0);
  printEq(slope,x0,y0);
  TLine* nL = makeLine(x0,x1,y0,y1,kRed,1); nL->Draw("same");

  // Testing
  TF1* ftop = new TF1("top","-0.57735026919*(x-0.3) + 0.173205080757",-1,1);
  ftop->Draw("same");
    
}

void printEq(double slope, double x, double y)
{
  cout<<"Eq: "<<slope<<"*(x-"<<x<<") + "<<y<<endl;
}

double rotation(double angle, double &x, double &y)
{
  
  double prevX = x;
  double prevY = y;

  x = cos(angle)*prevX - sin(angle)*prevY;
  y = sin(angle)*prevX + cos(angle)*prevY;
  

}


ray.py

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=#
# A class for the definition of a ray, which will be a vpython object  #
# called a curve. It is convenient for specifying a list of points and #
# having a line connecting them.  Ideal for ray-tracing purposes.      #
#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=#

from visual import curve

class ray:

    #----------------------------------#
    # Constructor
    #----------------------------------#
    def __init__(self,angle,x0,y0):
        
        # Set initial angle, x0, and y0
        self.angle = angle
        self.y     = y0
        self.x     = x0
        self.points = []
        
        self.addPoint(x0,y0)

    #----------------------------------#
    # Add method to update angle
    #----------------------------------#
    def update(self,angle,y,x):
        self.angle = angle
        self.y     = y
        self.x     = x

    #----------------------------------#
    # Add points
    #----------------------------------#
    def addPoint(self,x,y):
        self.points.append((x,y))
        

    #----------------------------------#
    # Get vector of last two points
    #----------------------------------#
    def getVector(self):

        np = len(self.points)
        if np < 2: 
            print "\t\t\tWarning: Not enough points..."
            return (0,0)

        p0 = self.points[np-2]
        p1 = self.points[np-1]
        #print "\t Making Vector: ", p1, p0
        newX = p1[0] - p0[0]
        newY = p1[1] - p0[1]
        vect = (newX, newY)
        return vect

    #----------------------------------#
    # Make curve object
    #----------------------------------#
    def drawRay(self):
        #print "Drawing Ray"
        #print self.points
        self.ray = curve(pos=self.points,
                         color=(0,0,0),
                         #radius=0.005
                         radius=0.002
                         )
        


cube.py

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-#
# This will be the virtual ice block and I wanted to have it as a class #
# so I can have methods to tell if the ray is still in the ice block.   #
# TODO: Need to fix the getTop, etc and also inCube methods. They are 
# broken when you perform rotations.
#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-#

from visual import *
import sys

class cube:
    

    #----------------------------------------#
    # Constructor
    #----------------------------------------#
    def __init__(self,x,y,z,L,H,Rotation):

        print "Making cube: ", x,y,z,L,H,Rotation

        # Save generic characteristics
        self.xpos     = x
        self.ypos     = y 
        self.zpos     = z
        self.length   = L
        self.height   = H
        self.rotation = Rotation

        # Create the box
        self.box = box(pos=(self.xpos,self.ypos,self.zpos),
                       length=L,
                       height=H,
                       width=0,
                       color=(0,1.5,1),
                       axis=(L*cos(Rotation),L*sin(Rotation),0))   

    #----------------------------------------#
    # Define method for rotation
    #----------------------------------------#
    def rotateCoords(self,x,y):
        newx = x*cos(self.rotation) - y*sin(self.rotation)
        newy = x*sin(self.rotation) + y*cos(self.rotation)
        return newx, newy

    #----------------------------------------#
    # Wrapper method to return the equations
    # the normals and the side numbers
    #----------------------------------------#
    def getSideInformation(self, initAngle):

        # Get the right to left results
        eqs   = self.getEquations()
        norms = self.getNormal()
        nums  = [0,1,2,3]

        # Rotate angle to be normal to 
        # the rotated cube
        angle = initAngle + self.rotation

        newNums = nums
        #if 0 <= angle and angle <= 90:
        #    newNums = nums
        if 90 < angle and angle <= 90+45:
            newNums  = [0,3,2,1]
        elif 90+45 < angle:
            newNums = [3,2,1,0]
        
        # Set the return values
        newEqs   = []
        newNorms = []
        for i in newNums:
            newEqs.append(eqs[i])
            newNorms.append(norms[i])
        
        # Return
        return newEqs, newNorms, newNums

    #----------------------------------------#    
    # Setup the equations
    #----------------------------------------#
    def getTop(self):   return self.buildEquation(0)
    def getRight(self): return self.buildEquation(1)
    def getBot(self):   return self.buildEquation(2)
    def getLeft(self):  return self.buildEquation(3)

    def getEquations(self):

        equations = []
        equations.append( self.getTop() )
        equations.append( self.getRight() )
        equations.append( self.getBot() )
        equations.append( self.getLeft() )
        #self.printEquations(equations)
        return equations

    def printEquations(self,eqs):
        for eq in eqs:
            s = eq[0]
            y0 = eq[1]
            x0 = eq[2]
            print "Eq: "+ str(s) + "*(x-"+str(x0)+") + "+str(y0) 
    #----------------------------------------#
    # Get normal vector for each side
    #----------------------------------------#
    def getTopV(self):
        return (-sin(self.rotation),cos(self.rotation))
    def getBotV(self):
        return (sin(self.rotation),-cos(self.rotation))
    def getRightV(self):
        return (cos(self.rotation),sin(self.rotation))
    def getLeftV(self):
        return (-cos(self.rotation), -sin(self.rotation))

    def getNormal(self):
        
        norms = []
        norms.append( self.getTopV() )
        norms.append( self.getRightV() )
        norms.append( self.getBotV() )
        norms.append( self.getLeftV() )
        return norms

    #----------------------------------------#
    # Determine if point is inside square
    # NOTE! This fails for angle > 90 deg
    #----------------------------------------#
    def inCube(self, point):
        x = point[0]
        y = point[1]

        # Get Equations
        Eqs = self.getEquations()
        if self.pointFails(0,Eqs[0],x,y): return False
        if self.pointFails(1,Eqs[1],x,y): return False
        if self.pointFails(2,Eqs[2],x,y): return False
        if self.pointFails(3,Eqs[3],x,y): return False

        return True



    #----------------------------------------#
    # Check if point passes
    #----------------------------------------#
    def pointFails(self,sideNum,eq,x,y):
        slope = eq[0]
        x0    = eq[2]
        y0    = eq[1]
        if slope == sys.float_info.max:
            #print "\t\t\tInfinite slope: ", x, x0
            if x0 < 0 and x < x0: return True
            if x0 > 0 and x > x0: return True
            return False
    
        ycomp = slope * (x-x0) + y0
        if sideNum == 0 and y > ycomp: return True
        if sideNum == 2 and y < ycomp: return True
        if sideNum == 1:
            if slope < 0 and y > ycomp: return True
            if slope > 0 and y < ycomp: return True
        if sideNum == 3:
            if slope < 0 and y < ycomp: return True
            if slope > 0 and y > ycomp: return True
        
        return False


    #----------------------------------------#
    # Get the slope, y0, and x0 position
    #----------------------------------------#
    def buildEquation(self,sideNum):

        # Deal with two specific cases: 0 and 90 degree rotation
        if self.rotation == 0: # Perfect cube
            if sideNum == 1 or sideNum == 3:
                x0 = self.length/2.
                if(sideNum == 3): x0 = -x0
                return (sys.float_info.max, 0, x0)
            else:
                y0 = self.height/2.
                if sideNum == 2: y0 = -y0
                return (0, y0, 0)
        elif self.rotation == 90*pi/180: # rotated cube
            if sideNum == 1 or sideNum == 3:
                y0 = -self.length/2.
                if sideNum == 3: y0 = -y0
                return (0,y0,0)
            else:
                x0 = self.height/2.
                if sideNum == 2: x0 = -x0
                return (sys.float_info.max, 0, x0)
        
        # Now handle generic rotation case
        x1 = 0
        x0 = 0
        y1 = 0
        y0 = 0
        
        # Start from a perfect cube
        if sideNum == 0:
            x0 = -self.length/2.
            x1 = self.length/2.
            y0 = self.height/2.
            y1 = self.height/2.
        elif sideNum == 1:
            x0 = self.length/2.
            x1 = self.length/2.
            y0 = -self.height/2.
            y1 = self.height/2.
        elif sideNum == 2:
            x0 = -self.length/2.
            x1 = self.length/2.
            y0 = -self.height/2.
            y1 = -self.height/2.
        elif sideNum == 3:
            x0 = -self.length/2.
            x1 = -self.length/2.
            y0 = -self.height/2.
            y1 = self.height/2.

        # Now add rotation
        x1,y1 = self.rotateCoords(x1,y1)
        x0,y0 = self.rotateCoords(x0,y0)
        
        # Define slope
        slope = (y1-y0)/(x1-x0)
        
        # Now return equation
        return (slope,y0,x0)


physics.py

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-#
# In here I will either put a list of methods or make it a class. The #
# idea will be to input some angle and get back either a refracted    #
# angle or the reflected angle.                                       #
#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-#

from math import *
import sys

#---------------------------------------------#
# Define constants needed for calculations
#---------------------------------------------#

# Index of refraction
nIce = 1.78
nAir = 1.

#---------------------------------------------#
# Define method to solve for crossing point
#---------------------------------------------#
def interactionPoint(rayAngle, rayY0, rayX0,
                     boundarySlope, boundaryY0, 
                     boundaryX0):

    # Check for non-zero slope
    # Which implies lines never cross
    # Also check for parallel lines (never cross)
    if boundarySlope == 0 and tan(rayAngle) == 0:
        print "\tInt point info: ", boundarySlope, rayAngle, tan(rayAngle)
        return (-9999,-9999)
    if boundarySlope == tan(rayAngle):
        return (-9999,9999)
    
    # Handle the case of x=constant
    if boundarySlope == sys.float_info.max:
        x = boundaryX0
        y = tan(rayAngle)*(x-rayX0) + rayY0
        #print "\t\tFixed x: ", x, y
        return (x,y)
    
    # All other cases can follow this:
    #print tan(rayAngle), boundarySlope
    x = 1/(tan(rayAngle) - boundarySlope) * (boundaryY0 - rayY0  + tan(rayAngle)*rayX0 - boundarySlope*boundaryX0)
    y = boundarySlope * (x-boundaryX0) + boundaryY0
    return (x,y)

#---------------------------------------------#
# Find incident angle
#---------------------------------------------#
def incidentAngle(v_ray, v_boundary,activeSide):
    
    # Angle we are after is the incident angle
    # of the ray on the boundary so we can
    # later check snell's law. 
    # Incident Angle = 90 - angle between vectors
    mag = v_ray[0]*v_boundary[0] + v_ray[1]*v_boundary[1]
    magRay = sqrt(v_ray[0]*v_ray[0]+v_ray[1]*v_ray[1])
    magB   = sqrt(v_boundary[0]*v_boundary[0]+v_boundary[1]*v_boundary[1])
    if mag/(magRay*magB) <= 1:
        return acos( mag/(magRay*magB) )
    else:
        return 0


#---------------------------------------------#
# Check for refraction
#---------------------------------------------#
def refractedAngle(incidentAngle, slope, activeSide, iceTilt):
    
    # Check for TIR case first
    arg = nIce/nAir * sin(fabs(incidentAngle))
    
    # If arg > 1, we have TIR
    if arg > 1: return -1

    # Otherwise, the ray leaves
    angle = asin(arg) 
    return angle


#---------------------------------------------#
# Get reflected angle
#---------------------------------------------#
def reflectedAngle(x0,y0,intX0,intY0,
                   incAngle, activeSide, rotAng):
    
    # I keep coming up with special cases in my code
    # when I think this should be handled very generically
    # So for now I will handle each side separately based
    # on the incident angle, the rotation angle and the active 
    # side
    
    # The rotation angle is actually for the cube
    # in the frames orientation.  We want to rotate
    # our coordinate system such that it is normal to the cube
    rotAng = -rotAng

    # The idea is to put the coordinate system at
    # the interaction point with normal to the side
    # in the positive y direction and then check where
    # our initial ray point lies to determine the angle
    # in the coordinate system of the block
    xt = x0 - intX0  # translate x
    yt = y0 - intY0  # translate y
    xr = xt*cos(rotAng) - yt*sin(rotAng)
    yr = xt*sin(rotAng) + yt*cos(rotAng)

    # Top of cube
    pi2 = pi/2.

    if activeSide == 0: 
        if xr > 0: return pi2 - incAngle
        return pi2 + incAngle
    # Right
    elif activeSide == 1:
        if yr > 0: return -pi + incAngle
        return pi - incAngle
    # Bottom
    elif activeSide == 2:
        if xr < 0: return pi2 - incAngle
        return pi2 + incAngle
    # Left
    elif activeSide == 3:
        if yr < 0: return incAngle
        return -incAngle

    # Do nothing
    return refAng

#---------------------------------------------#
# Translate angle to det coords
#---------------------------------------------#
def translateAngle(x0, y0, intX0, intY0, 
                   rotAng, refAng, activeSide):
    # x0,y0 = original position
    # rayX0,rayY0 = incident point on wall
    # refAngle = refracted angle
    # activeSide = side of the cube (0=top, 1=right, etc)
    
    # The rotation angle is actually for the cube
    # in the frames orientation.  We want to rotate
    # our coordinate system such that it is normal to the cube
    rotAng = -rotAng

    # The idea is to put the coordinate system at
    # the interaction point with normal to the side
    # in the positive y direction and then check where
    # our initial ray point lies to determine the angle
    # in the coordinate system of the block
    xt = x0 - intX0  # translate x
    yt = y0 - intY0  # translate y
    xr = xt*cos(rotAng) - yt*sin(rotAng)
    yr = xt*sin(rotAng) + yt*cos(rotAng)

    #print xt, yt, xr, yr, rotAng*180./pi

    # Remember, refracted angle is w/respect to
    # the normal.  Make it with respect to x-axis

    # Top of cube
    pi2 = pi/2.
    if activeSide == 0: 
        if xr > 0: return pi2 + refAng
        return pi2 - refAng
    # Right
    elif activeSide == 1:
        if yr > 0: return -refAng
        return refAng
    # Bottom
    elif activeSide == 2:
        if xr < 0: return refAng - pi2
        return -refAng - pi2
    # Left
    elif activeSide == 3:
        if yr < 0: return pi-refAng 
        return pi+refAng

    # Do nothing
    return refAng

#---------------------------------------------#
# Get perpendicular vector
#---------------------------------------------#
def getPerp(point):
    x = point[0]
    y = point[1]
    
    return (y,x)


main.py
#!/usr/local/bin/python

# This is for runnning on linux
#!/usr/bin/python

# For running on my mac
#!/usr/local/bin/python

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=#
# This will be the 'executable' that will draw the ray tracing for the ice #
# block. Will add some options in the future, but at first we consider a   #
# simple case.                                                             #
#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=#

#-------------------------------------#
# Import necessaries
#-------------------------------------#

from visual import *
from cube import *
from ray import *
from physics import *
import sys
from optparse import OptionParser

#-------------------------------------#
# Parse some simple user options
#-------------------------------------#

#
## Setup parser
#
parser = OptionParser()

parser.add_option("-s", "--stepsize", action="store",
                  #type=int, default=10, dest="stepsize",
                  type=float, default=10, dest="stepsize",
                  help="Option to set number of steps")
parser.add_option("-l", "--less90", action="store_true",
                  default=False, dest="L90",
                  help="Option to run angles 0-90")
parser.add_option("-g", "--greater90", action="store_true",
                  default=False, dest="G90",
                  help="Optin to run angles 90-180")
parser.add_option("-a","--angle",action="store",
                  type=float,default=-1,dest="angle",
                  help="Specify a specific angle")
parser.add_option("-r", "--rotation", action="store",
                  type=float, default=-30,dest="rotAngle",
                  help="Specify rotation angle for ice")
parser.add_option("-i", "--injection", action="store",
                  type=float, default=0,dest="injPoint",
                  help="Specify injection point in ice")

#
## Load the options
#

print "-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-"
print "Starting ray tracing program"
print ""

opts, args = parser.parse_args()

angles = []
stepsize = opts.stepsize
if opts.L90:
    angMax = 90
    angle  = 0
    while angle <= angMax:
        angles.append( angle )
        angle += stepsize
    print "Running for angles 0-90"
elif opts.G90:
    angMax = 180
    angle  = 90
    while angle <= angMax:
        angles.append( angle )
        angle += stepsize
    print "Running for angles 90-180"
elif opts.angle > 0:
    angles.append( opts.angle )
    print "Running for angle: ", angles[0]
else:
    print "Please specify angles to use."
    print ""
    sys.exit()

print angles
print ""
print ""


#
## Convert all angles to radians
#
for i in range(len(angles)):
    angles[i] = angles[i] * pi/180.

#
## Save rotation angle
#

rotAng = opts.rotAngle * pi/180

#-------------------------------------#
# Setup the environment
#-------------------------------------#

m_world = 1
window = display(title='Ray Tracing', x=0, y=0, 
                 width=800, height=800,
                 center=(0,0,0), 
                 #center=(5,0,0), 
                 #scale=(0.1,0.1,1),
                 background=(1,1,1),                 
                 range=m_world)

#-------------------------------------#
# Add antenna Pole off to the right
#-------------------------------------#

#p_height = 8 # [m]
#p_center = (8,-0.15,0)  # [m]
#pole = cylinder(pos = p_center,
#                axis = (0,p_height,0),
#                radius = 0.1,
#                color=(1,0,0))


#-------------------------------------#
# Make the ice block
#-------------------------------------#

# Dimensions
length  = 1  # [m]
#length  = 0.5  # [m]
#length  = 0.3  # [m]
height  = 0.3 # [m]
width   = 0.3 # [m] This is not used. Just 2D

# Locations
icex = 0
icey = 0
icez = 0

# make the cube
iceblock = cube(icex,icey,icez,length,height,rotAng)

# Draw axis
#yaxis = curve(pos=[(0,height),(0,-height)],color=(0,0,0),radius = -0.003)
#xaxis = curve(pos=[(length,0),(-length,0)],color = (0,0,0),radius = -0.003)

#-------------------------------------#
# Determine which angles to draw
#-------------------------------------#

# Specify x0 and y0 for beam
botEq = iceblock.getBot()
x0    = opts.injPoint
y0    = botEq[0]*(x0-botEq[2]) + botEq[1] + 0.06

# Add Rays for each angle and calculate
# the points
rays = []
for ang in angles:

    # Get equations defining cube
    #iceEq    = iceblock.getEquations()
    #normals  = iceblock.getNormal()
    #sideNums = [0,1,2,3]
    iceEq, normals, sideNums = iceblock.getSideInformation(ang*180/pi)

    # Hack for angles > 90
    #cornerAngle = pi - atan((iceblock.height/2.-y0)/(iceblock.length/2.-x0)) + iceblock.rotation
    #if ang > cornerAngle:
    #    temp = iceEq 
        #iceEq = [temp[3], temp[2], temp[1], temp[0]]
    #    iceEq = [temp[0], temp[3], temp[2], temp[1]]
    #    temp = normals
        #normals = [temp[3], temp[2], temp[1], temp[0]]
    #    normals = [temp[0], temp[3], temp[2], temp[1]]
        #sideNums = [3,2,1,0]
    #    sideNums = [0,3,2,1]
        

    # Print some info
    print "------------------------------------------------------------"
    print "Trying angle: ", ang, ang*180/pi

    # Create new ray for this angle
    newray = ray(ang,x0,y0)

    # Loop until refracted angle is positive
    # or hits all four sides
    angRef   = -1
    nSides   = 0
    activeSide = -1
    point = -1
    while angRef < 0 and nSides <= 10:

        # Get interaction point
        intPoint = (0,0)
        v_normal   = (0,0)
        insideIce = False
        for i in range(len(iceEq)):
            eq = iceEq[i]
            #print "Checking interaction point: ", sideNums[i]
            intPoint = interactionPoint(newray.angle,
                                        newray.y,
                                        newray.x,
                                        eq[0],
                                        eq[1],
                                        eq[2])

            if iceblock.inCube(intPoint): 
                insideIce  = True
                activeSide = sideNums[i]
                point      = i
                break

        # Change order of iceEq for next iteration
        # in case of reflection
        saveEq = iceEq[point]
        del iceEq[point]
        iceEq.append(saveEq)
        v_normal = normals[point]
        del normals[point]
        normals.append(v_normal)
        del sideNums[point]
        sideNums.append(activeSide)

        # Protect against exhausting all options
        if not insideIce: break
        print "\t\tIncident Point: ",intPoint, " on side", activeSide
        newray.addPoint(intPoint[0],intPoint[1])

        # Get Incident angle
        incAngle = incidentAngle(newray.getVector(),v_normal,activeSide)
        print "\t\tIncident Angle: ", incAngle*180/pi

        # Check if there is refraction
        angRef = refractedAngle(incAngle, newray.angle, activeSide, rotAng)

        if angRef < 0: 
            print "\t\t\tTIR!"
            newAngle = reflectedAngle(newray.x,newray.y,intPoint[0],intPoint[1],
                                      incAngle,activeSide,rotAng)
            newAngle += iceblock.rotation
            print "\t\t\tReflected Angle", (pi/2 - incAngle)*180/pi
            print "\t\t\tReflected Angle in frame coords", newAngle*180/pi 
            newray.update(newAngle,intPoint[1],intPoint[0])
            #newray.addPoint(intPoint[0], intPoint[1])
        else:
            print "\t\t\tRefracted Angle", angRef*180/pi
            fixedAngle = translateAngle(newray.x,newray.y,
                                        intPoint[0],intPoint[1],
                                        rotAng,angRef,activeSide)
            print "\t\t\t\tTranslated Angle", fixedAngle*180/pi
            fixedAngle += iceblock.rotation
            print "\t\t\t\tIncluding Rotation", fixedAngle*180/pi
            x = 0
            y = 0
            sign = 1

            x = m_world*cos(fixedAngle)
            y = m_world*sin(fixedAngle)
            
            # Add final point off in the distance
            newray.addPoint(intPoint[0]+x,intPoint[1]+y)

        # Incraease nSides hit by one
        nSides += 1

    # end while loop

    newray.drawRay()

#end loop over angles
sys.exit()
